### DISCUSSION TITLE
Mark unused variables in default/main.yml or vars/main.yml

### DISCUSSION THREAD

**Discussion by maunzCache** [ansible/ansible-lint#1301] [Ideas] [answered]
In my company we have +50 Ansible Roles and at least 50 Ansible Playbooks which require management regarding Ansible compatibility, Python updates, documentation and so on. We managed to find solutions for a few of the problems arising when using so many roles. One particular issue we cannot seem to fix is the usage of declared but unused variables in roles.

E.g.
A developer can declare variables in either default/main.yml or vars/main.yml which are imported in most cases. Thus those variables can be used anytime in handlers, tasks and templates or encapsuled in other vars.

I think ansible-lint is a perfect fit to check whether declared variables have a reference in a role. Maybe the support for this kind linting should be limited as some variables may only exist due to local facts or other means. Those facts existing at runtime should be out of scope for this idea. Also local scope variables which are created through set_facts or register can be ignored for this idea. There may be support for local variables (facts) in single tasks and/or imported or included facts.

**Comment by ssbarnea** [comment, marked as answer]
Not possible.

**Reply by maunzCache** [reply]
I don't want to be picky but is there a reason why it is not possible?

**Reply by phemmer** [reply]
Limitations of static code analysis. The variable may be referenced inside a Jinja template, or inside a module. This would require ansible-lint to have the ability to parse the 2 languages, which would be incredibly difficult. Even within a normal yaml playbook, it would have to differentiate between a string that happens to contain the variable's name, and something referencing the variable itself, which isn't possible as its up to the module how its parameters are interpreted.

**Reply by hoo29** [reply]
I've created [little-timmy](https://pypi.org/project/little-timmy/), a cli tool that can identify many types of unused variables in playbooks and roles. While it's not perfect, it has been a significant help for us in managing our playbooks. Hopefully, it can be useful to others as well!
